### Basic working of Google Palm LLM in LangChain

In [15]:
from langchain.llms import GooglePalm

api_key = 'AIzaSyAaBtFel2BHQqLMZjHxLKpvcbnKn14zz8E' 

llm = GooglePalm(google_api_key=api_key, temperature=0.1)

In [16]:
poem = llm("Write a 4 line poem of my love for shiv")
print(poem)

Shiv, my lord, my master,
Your love is like a river,
Flowing through my heart,
Embracing me with its warmth.


In [17]:
essay = llm("write praise for electronic item")
print(essay)

**Praise for the Amazon Kindle Paperwhite**

The Amazon Kindle Paperwhite is a truly remarkable electronic device. It is lightweight, portable, and has a long battery life. The display is also stunning, with a high resolution and a warm light that makes it easy to read for extended periods of time.

I have been using the Kindle Paperwhite for several months now, and I have been very impressed with its performance. I love being able to carry my entire library with me wherever I go, and I appreciate the fact that the Kindle is so easy to use. The interface is simple and intuitive, and the device is very responsive.

I would highly recommend the Amazon Kindle Paperwhite to anyone who loves to read. It is an excellent e-reader that is sure to provide you with years of enjoyment.

Here are some specific things I love about the Kindle Paperwhite:

* The display is beautiful. The text is sharp and clear, and the colors are vibrant.
* The battery life is excellent. I can easily go for weeks wi

In [18]:
from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Techstack and roadmap FAQ csv file

In [19]:
from langchain.document_loaders.csv_loader import CSVLoader

loader = CSVLoader(file_path='tech_faqs.csv', source_column="prompt")

# Store the loaded data in the 'data' variable
data = loader.load()
data

[Document(page_content="prompt: What's the best programming language for Data Structures and Algorithms (DSA)?\nresponse: The choice of the best programming language for DSA often comes down to personal preference, but popular languages include Python, Java, and C++. Python is a great choice for beginners because of its simplicity and readability, while Java and C++ offer strong performance and are commonly used for competitive programming.\n: The choice of the best programming language for DSA often comes down to personal preference, but popular languages include Python, Java, and C++. Python is a great choice for beginners because of its simplicity and readability, while Java and C++ offer strong performance and are commonly used for competitive programming.", metadata={'source': "What's the best programming language for Data Structures and Algorithms (DSA)?", 'row': 0}),
 Document(page_content='prompt: C++ vs. Java: Which one should I learn for competitive programming and software d

### Hugging Face Embeddings 

In [20]:
!pip install transformers

In [21]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

embeddings = HuggingFaceInstructEmbeddings(
    query_instruction="Represent the query for retrieval: "
)
e = embeddings.embed_query("How can I contact the instructors for any doubts/support?")

load INSTRUCTOR_Transformer
max_seq_length  512


In [22]:
len(e)

768

In [23]:
e[:5]

[0.005723217967897654,
 -0.0028625724371522665,
 -0.021918946877121925,
 0.0028228696901351213,
 0.04676933214068413]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [24]:
from langchain.embeddings import HuggingFaceInstructEmbeddings
from langchain.vectorstores import FAISS

instructor_embeddings =  HuggingFaceInstructEmbeddings()
# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)
# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)


load INSTRUCTOR_Transformer
max_seq_length  512


In [25]:
retriever = vectordb.as_retriever()
rdocs = retriever.get_relevant_documents("c++ vs java?")
rdocs

[Document(page_content='prompt: C++ vs. Java: Which one should I learn for competitive programming and software development?\nresponse: Both C++ and Java are powerful languages. C++ is preferred by many competitive programmers due to its efficiency and access to data structures in the C++ Standard Template Library (STL). Java, on the other hand, is widely used in enterprise software development and Android app development. Choose based on your specific interests and goals.\n: ', metadata={'source': 'C++ vs. Java: Which one should I learn for competitive programming and software development?', 'row': 1}),
 Document(page_content="prompt: What's the best programming language for Data Structures and Algorithms (DSA)?\nresponse: The choice of the best programming language for DSA often comes down to personal preference, but popular languages include Python, Java, and C++. Python is a great choice for beginners because of its simplicity and readability, while Java and C++ offer strong perfor

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [26]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain along with prompt template 🚀

In [27]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


### We are all set 👍🏼 Let's ask some questions now

In [28]:
chain('best way to start programming solving interviews?')

{'query': 'best way to start programming solving interviews?',
 'result': 'Focus on practicing coding problems regularly on platforms like LeetCode, HackerRank, or CodeSignal. Join online coding communities, participate in mock interviews, and study data structures and algorithms to excel in technical interviews.',
 'source_documents': [Document(page_content='prompt: How can I improve my problem-solving skills for coding interviews?\nresponse: Focus on practicing coding problems regularly on platforms like LeetCode, HackerRank, or CodeSignal. Join online coding communities, participate in mock interviews, and study data structures and algorithms to excel in technical interviews.\n: ', metadata={'source': 'How can I improve my problem-solving skills for coding interviews?', 'row': 4}),
  Document(page_content='prompt: What is the best way to start learning programming for a complete beginner?\nresponse: Start with an easy-to-understand language like Python. Utilize online resources, tut

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [29]:
chain("Devops?")

{'query': 'Devops?',
 'result': '\n response: DevOps is a software development methodology that emphasizes the collaboration between software developers and IT operations teams. The goal of DevOps is to shorten the development lifecycle and improve the quality of software products by automating processes and breaking down silos between teams.\n: ',
 'source_documents': [Document(page_content='prompt: is hackcbs a offline hackathon?\nresponse: Yes\n: ', metadata={'source': 'is hackcbs a offline hackathon?', 'row': 5}),
  Document(page_content='prompt: C++ vs. Java: Which one should I learn for competitive programming and software development?\nresponse: Both C++ and Java are powerful languages. C++ is preferred by many competitive programmers due to its efficiency and access to data structures in the C++ Standard Template Library (STL). Java, on the other hand, is widely used in enterprise software development and Android app development. Choose based on your specific interests and goal

In [30]:
chain("Java course?")

{'query': 'Java course?',
 'result': '\n response: Learn Java from scratch with this comprehensive course. Learn the fundamentals of Java programming, including variables, data types, operators, control flow statements, methods, classes, and objects.\n: Learn Java from scratch with this comprehensive course. Learn the fundamentals of Java programming, including variables, data types, operators, control flow statements, methods, classes, and objects.',
 'source_documents': [Document(page_content='prompt: C++ vs. Java: Which one should I learn for competitive programming and software development?\nresponse: Both C++ and Java are powerful languages. C++ is preferred by many competitive programmers due to its efficiency and access to data structures in the C++ Standard Template Library (STL). Java, on the other hand, is widely used in enterprise software development and Android app development. Choose based on your specific interests and goals.\n: ', metadata={'source': 'C++ vs. Java: Whic

In [31]:
chain("Do you have plans to launch blockchain course in future?")

{'query': 'Do you have plans to launch blockchain course in future?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='prompt: is hackcbs a offline hackathon?\nresponse: Yes\n: ', metadata={'source': 'is hackcbs a offline hackathon?', 'row': 5}),
  Document(page_content='prompt: What is the best way to start learning programming for a complete beginner?\nresponse: Start with an easy-to-understand language like Python. Utilize online resources, tutorials, and courses like the one offered by Codebasics to build a strong foundation. Practice coding regularly and gradually move to more advanced topics.\n: ', metadata={'source': 'What is the best way to start learning programming for a complete beginner?', 'row': 3}),
  Document(page_content='prompt: How do I choose a specialization in computer science?\nresponse: Choosing a specialization depends on your interests and career goals. Options include software development, data science, machine learning, web development

In [32]:
chain("should I learn power bi or tableau?")

{'query': 'should I learn power bi or tableau?',
 'result': 'Both Power BI and Tableau are powerful business intelligence tools that can help you visualize and analyze data. However, there are some key differences between the two platforms that you should consider before making a decision.\n\nPower BI is a cloud-based platform that is integrated with Microsoft 365. This makes it easy to share reports and dashboards with colleagues and collaborate on projects. Tableau, on the other hand, is a self-hosted platform that can be installed on-premises or in the cloud. This gives you more control over your data and security, but it also requires more technical expertise to set up and manage.\n\nIn terms of features, Power BI and Tableau are both very similar. Both platforms offer a wide range of data visualization options, including charts, graphs, maps, and dashboards. They also both support data blending, which allows you to combine data from multiple sources into a single report.\n\nUltima

In [33]:
chain("I've a MAC computer. Can I use powerbi on it?")

{'query': "I've a MAC computer. Can I use powerbi on it?",
 'result': 'Yes, you can use Power BI on a Mac.\n: Yes, you can use Power BI on a Mac.',
 'source_documents': [Document(page_content='prompt: is hackcbs a offline hackathon?\nresponse: Yes\n: ', metadata={'source': 'is hackcbs a offline hackathon?', 'row': 5}),
  Document(page_content='prompt: Nearest station To Hackcbs\nresponse: Rithala\n: ', metadata={'source': 'Nearest station To Hackcbs', 'row': 6}),
  Document(page_content="prompt: What's the best programming language for Data Structures and Algorithms (DSA)?\nresponse: The choice of the best programming language for DSA often comes down to personal preference, but popular languages include Python, Java, and C++. Python is a great choice for beginners because of its simplicity and readability, while Java and C++ offer strong performance and are commonly used for competitive programming.\n: The choice of the best programming language for DSA often comes down to personal pr

In [34]:
chain("I don't see power pivot. how can I enable it?")

{'query': "I don't see power pivot. how can I enable it?",
 'result': 'To enable Power Pivot, follow these steps:\n1. Open Excel.\n2. Click the File tab.\n3. Click Options.\n4. Click Add-ins.\n5. Click Manage: COM Add-ins.\n6. Click Go.\n7. Select Microsoft Power Pivot for Excel 2016.\n8. Click OK.\n9. Click OK.\n10. Restart Excel.\n: ',
 'source_documents': [Document(page_content='prompt: Nearest station To Hackcbs\nresponse: Rithala\n: ', metadata={'source': 'Nearest station To Hackcbs', 'row': 6}),
  Document(page_content='prompt: is hackcbs a offline hackathon?\nresponse: Yes\n: ', metadata={'source': 'is hackcbs a offline hackathon?', 'row': 5}),
  Document(page_content='prompt: How do I choose a specialization in computer science?\nresponse: Choosing a specialization depends on your interests and career goals. Options include software development, data science, machine learning, web development, cybersecurity, and more. Research each field, consider your interests, and explore in

In [35]:
chain("What is the price of your machine learning course?")

{'query': 'What is the price of your machine learning course?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='prompt: How do I choose a specialization in computer science?\nresponse: Choosing a specialization depends on your interests and career goals. Options include software development, data science, machine learning, web development, cybersecurity, and more. Research each field, consider your interests, and explore introductory courses to find the right fit.\n: ', metadata={'source': 'How do I choose a specialization in computer science?', 'row': 2}),
  Document(page_content="prompt: What's the best programming language for Data Structures and Algorithms (DSA)?\nresponse: The choice of the best programming language for DSA often comes down to personal preference, but popular languages include Python, Java, and C++. Python is a great choice for beginners because of its simplicity and readability, while Java and C++ offer strong performance and are commonly